src/Core/Managed/Shared/Extensibility/Metrics/MetricExtractorTelemetryProcessorBase.cs

```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using System.Globalization;
+
+using Microsoft.ApplicationInsights.Channel;
+using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+namespace Microsoft.ApplicationInsights.Extensibility.Metrics
+{
+    public abstract class MetricExtractorTelemetryProcessorBase : ITelemetryProcessor, ITelemetryModule, IDisposable
+    {
+        private MetricManager _metricManager = null;
```

Comment:
Ok, so to continue my response above - how do I chose between the two options.

* Making thing configurable publicly "just for ourselves" is not a good approach. It creates a lot of work and hardening requirements. So the extractors should be either not configurable at all, or - if they appear in ai.config - then customers should be able to write their own. As mentioned, based on Sergey's recommendation, I am going with the appear in ai.config route.

So, not I want to make it super easy to write an extractor. So I put common logic in a common base. But a key engineering principle is to avoid any unnecessary abstractions. So to test whether my base is factored correctly i should design to have at least two specializations. So instead of having just one specialization for "all auto-collected stuff", i am going with 2 specializations for "all request related stuff" and "all dependency related stuff". And additional benefit of this approach is that if in the future we add a package to collect some kind of SpecialTelemetry we will be able to have that package add SpecialTelemetryMetricExtractor using a consistent method.